demos/greengrass_connectivity/aws_greengrass_discovery_demo.c

```diff
@@ -85,129 +145,344 @@ typedef struct
+                }
+            }
+        }
+    } while ( ( xNetworkStatus != TRANSPORT_SOCKET_STATUS_SUCCESS || xMqttStatus != pdPASS ) && ( xBackoffStatus == pdPASS ) );
+
+    return ( xNetworkStatus == TRANSPORT_SOCKET_STATUS_SUCCESS && xMqttStatus == pdPASS ) ? pdPASS : pdFAIL;
+}
+
+static BaseType_t prvMQTTPublishToTopic( MQTTContext_t * pxMQTTContext )
+{
+    MQTTStatus_t xResult;
+    MQTTPublishInfo_t xMQTTPublishInfo;
+    BaseType_t xStatus = pdPASS;
+    uint32_t ulMessageCounter;
+    char cBuffer[ ggdDEMO_MAX_MQTT_MSG_SIZE ];
+    const char * pcTopic = ggdDEMO_MQTT_MSG_TOPIC;
+
+    /* Some fields are not used by this demo so start with everything at 0. */
+    ( void ) memset( ( void * ) &xMQTTPublishInfo, 0x00, sizeof( xMQTTPublishInfo ) );
+
+    /* This demo uses QoS1. */
+    xMQTTPublishInfo.qos = MQTTQoS0;
+    xMQTTPublishInfo.retain = false;
+    xMQTTPublishInfo.pTopicName = pcTopic;
+    xMQTTPublishInfo.topicNameLength = ( uint16_t ) strlen( pcTopic );
+
+    for( ulMessageCounter = 0; ulMessageCounter < ( uint32_t ) ggdDEMO_MAX_MQTT_MESSAGES; ulMessageCounter++ )
+    {
+        xMQTTPublishInfo.pPayload = ( const void * ) cBuffer;
+        xMQTTPublishInfo.payloadLength = ( uint32_t ) sprintf( cBuffer, ggdDEMO_MQTT_MSG_DISCOVERY, ( long unsigned int ) ulMessageCounter ); /*lint !e586 sprintf can be used for specific demo. */
+
+        /* Get a unique packet id. */
+        usPublishPacketIdentifier = MQTT_GetPacketId( pxMQTTContext );
 
-        IotLogInfo( "Attempting to establish MQTT connection to Greengrass." );
-        xMqttStatus = IotMqtt_Connect( &xNetworkInfo,
-                                       &xMqttConnectInfo,
-                                       _maxCommandTimeMs,
-                                       pMqttConnection );
+        /* Send PUBLISH packet. Packet ID is not used for a QoS1 publish. */
+        xResult = MQTT_Publish( pxMQTTContext, &xMQTTPublishInfo, usPublishPacketIdentifier );
 
-        if( xMqttStatus != IOT_MQTT_NETWORK_ERROR )
+        if( xResult != MQTTSuccess )
         {
-            break;
+            xStatus = pdFAIL;
+            IotLogError( ( "Failed to send PUBLISH message to broker: Topic=%s, Error=%s",
+                           pcTopic,
+                           MQTT_Status_strerror( xResult ) ) );
         }
+
+        IotLogInfo( "MQTT PUBLISH message sent successfully to the broker: Topic=%s", pcTopic );
+        vTaskDelay( pdMS_TO_TICKS( _timeBetweenPublishMs ) );
     }
 
-    return xMqttStatus;
+    return xStatus;
 }
 
-/*-----------------------------------------------------------*/
+static void _sendMessageToGGC( MQTTContext_t * pxMQTTContext )
+{
+    BaseType_t xDemoStatus = pdFAIL;
+
+    /* Call the MQTT PUBLISH function. */
+    xDemoStatus = prvMQTTPublishToTopic( pxMQTTContext );
+
+    if( xDemoStatus != pdPASS )
+    {
+        IotLogError( "mqtt_client - Failure to publish." );
```

Comment:
Preferably we just use LogError function instead of IotLogError. Is there any reason why you decided to use IotLogError?